slf4j/src/main/java/com/google/common/flogger/backend/slf4j/Slf4jLoggerBackend.java

```diff
@@ -0,0 +1,188 @@
+   * <td>INFO</td>
+   * </tr>
+   * <tr>
+   * <td>WARNING</td>
+   * <td>WARN</td>
+   * </tr>
+   * <tr>
+   * <td>SEVERE</td>
+   * <td>ERROR</td>
+   * </tr>
+   * </table>
+   *
+   * <p>Custom JUL levels are mapped to the next-lowest standard JUL level; for example, a custom
+   * level at 750 (between INFO:800 and CONFIG:700) would map to the same as CONFIG (DEBUG).</p>
+   *
+   * <p>It isn't expected that the JUL levels 'ALL' or 'OFF' are passed into this method;
+   * doing so will throw an IllegalArgumentException, as those levels are for configuration, not
+   * logging</p>
+   *
+   * @param level the JUL level to map; any standard or custom JUL level, except for ALL or OFF
+   * @return the MappedLevel object representing the SLF4J adapters appropriate for the requested
+   * log level; never null.
+   */
+  private static Slf4jLogLevel mapToSlf4jLogLevel(Level level) {
+    int requestedLevel = level.intValue();
+
+    if (requestedLevel == Level.ALL.intValue() || requestedLevel == Level.OFF.intValue()) {
+      // Flogger doesn't allow ALL or OFF to be used for logging, only for configuration
+      throw new IllegalArgumentException("Unsupported log level: " + level);
+    }
+
+    // FINEST, FINER -> TRACE
+    // custom JUL levels less than FINE -> TRACE
+    if (requestedLevel < Level.FINE.intValue()) {
+      return Slf4jLogLevel.TRACE;
+    }
+
+    // FINE, CONFIG -> DEBUG
+    // custom JUL levels less than INFO -> DEBUG
+    if (requestedLevel < Level.INFO.intValue()) {
+      return Slf4jLogLevel.DEBUG;
+    }
+
+    // INFO -> INFO
+    // custom JUL levels less than WARNING -> INFO
+    if (requestedLevel < Level.WARNING.intValue()) {
+      return Slf4jLogLevel.INFO;
+    }
+
+    // WARNING -> WARN
+    // custom JUL levels less than SEVERE -> WARN
+    if (requestedLevel < Level.SEVERE.intValue()) {
+      return Slf4jLogLevel.WARN;
+    }
+
+    // SEVERE -> ERROR
+    // custom JUL levels greater than SEVERE -> ERROR
+    return Slf4jLogLevel.ERROR;
+  }
+
+  @Override
+  public String getLoggerName() {
+    return logger.getName();
+  }
+
+  @Override
+  public boolean isLoggable(Level level) {
+    Slf4jLogLevel slf4jLogLevel = mapToSlf4jLogLevel(level);
+
+    // dispatch to each level-specific method as SLF4J doesn't expose a logger.log( level, ... )
+    switch (slf4jLogLevel) {
+      case TRACE:
+        return logger.isTraceEnabled();
+      case DEBUG:
+        return logger.isDebugEnabled();
+      case INFO:
+        return logger.isInfoEnabled();
+      case WARN:
+        return logger.isWarnEnabled();
+      case ERROR:
+        return logger.isErrorEnabled();
+      default:
+        throw new AssertionError("Unknown SLF4J log level: " + slf4jLogLevel);
+    }
+  }
+
+  @Override
+  public void log(LogData data) {
+    SimpleMessageFormatter.format(data, this);
+  }
+
+  @Override
+  public void handleError(RuntimeException exception, LogData badData) {
+    // log at WARN to ensure visibility
```

Comment:
Technically WARN does not ensure visibility. I'd either go with ERROR here or at least make it ERROR if the original log statement was ERROR. I don't think we want to downgrade the level if something goes wrong.